File: 2405.1/jf_syn/security/WeakXmlTransformersSecurityChecker/WeakXmlTransformersSecurity.java
//WEAK_XML_TRANSFORMERS_SECURITY
package security;

import javax.servlet.http.HttpServletRequest;
import java.util.Hashtable;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;


public class WeakXmlTransformersSecurity {

    public void test() {
        //...        
        Transformer transformer = TransformerFactory.newInstance().newTransformer();    // @violation
        TransformerFactory factory = TransformerFactory.newInstance();
        Transformer transformer2 = factory.newTransformer();   // @violation

        TransformerFactory factory2 = TransformerFactory.newInstance();
        factory2.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory2.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
        Transformer transformer3 = factory2.newTransformer();    /* Safe */
        //...
    }

}

File: 2405.1/kotlinchk/security/WeakXmlTransformersSecurityChecker/WeakXmlTransformersSecurity.kt
package testcase.rule.security.WeakXmlTransformersSecurityChecker

import javax.xml.XMLConstants
import javax.xml.transform.Transformer
import javax.xml.transform.TransformerFactory

object TestCase {
	fun violation1() {
		val transformer = TransformerFactory.newInstance().newTransformer()
		transformer.transform(input, result)			// @violation
	}

	fun violation2() {
		val factory1 = TransformerFactory.newInstance()
		factory1.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
		factory1.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "")

		val factory2 = TransformerFactory.newInstance()

		val transformer = factory2.newTransformer()
		transformer.transform(input, result)			// @violation
	}

	fun violation3() {
		val factory = TransformerFactory.newInstance()
		factory.setAttribute(XMLConstants.XML_NS_URI, "")

		val transformer = factory.newTransformer()
		transformer.transform(input, result)			// @violation
	}

	fun good1() {
		val factory = TransformerFactory.newInstance()
		factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
		factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "")

		val transformer = factory.newTransformer()
		transformer.transform(input, result)			// good
	}

	fun good2() {
		val factory = TransformerFactory.newInstance()
		factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true)

		val transformer = factory.newTransformer()
		transformer.transform(input, result)			// good
	}
}